metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -892,48 +915,73 @@ class MigrationWriteOperation implements ControllerWriteOperation<Void> {
             }
             @Override
             public ControllerResult<Void> generateRecordsAndResult() {
-                return ControllerResult.atomicOf(batch, null);
+                return ControllerResult.of(batch, null);
             }
 
             public void processBatchEndOffset(long offset) {
                 highestMigrationRecordOffset = new OffsetAndEpoch(offset, curClaimEpoch);
             }
         }
         @Override
-        public void beginMigration() {
-            log.info("Starting ZK Migration");
-            // TODO use KIP-868 transaction
+        public CompletableFuture<?> beginMigration() {
+            if (featureControl.metadataVersion().isMetadataTransactionSupported()) {
+                log.info("Starting ZK Migration");
```

Comment:
How about a message like "Starting initial ZooKeeper state lift and shift process" ?